### PR TITLE
Add message when unrecognised parameter is passed to constellation

### PR DESCRIPTION
**Pull request by n-hutton**
This pr adds the following behaviour to constellation:

```
$ ./constellation -foobar
[3/3] Linking CXX executable apps/constellation/constellation
 F E ╱     Constellation v0.7.0-alpha3-4-ga7634adf
   T C     Copyright 2018-2019 (c) Fetch AI Ltd.
     H

[E] 2019/07/27 13:12:38 | SettingCollection              : Unrecognised parameter(s) passed.
[E] 2019/07/27 13:12:38 | SettingCollection              : Unrecognised parameter: -foobar
[I] 2019/07/27 13:12:38 | SettingCollection              :
Valid parameters:
-lanes                        "The number of lanes to be used"
-slices                       "The number of slices to be used"
-block-interval               "The block interval is milliseconds"
-standalone                   "Signal the network should run in standalone mode"
-private-network              "Signal the network should run as part of a private network"
-db-prefix                    "The block interval is milliseconds"
-port                         "The starting port for ledger services"
-peers                        "The comma separated list of addresses to initially connect to"
-external                     "This node's global IP address or hostname"
-config                       "The path to the manifest configuration"
-max-peers                    "The max number of peers to connect to"
-transient-peers              "The number of the peers which will be random in answer sent to peer requests"
-peers-update-cycle-ms        "How fast to do peering updates"
-disable-signing              "Disable the signing of all network messages"
-bootstrap                    "Signal that we should connect to the bootstrap server"
-discoverable                 "Signal that this node can be advertised on the bootstrap server"
-host-name                    "The hostname or identifier for this node"
-network                      "The name of the bootstrap network to connect to"
-token                        "The authentication token when talking to bootstrap"
-processor-threads            "The number of processor threads"
-verifier-threads             "The number of verifier threads"
-executors                    "The number of transaction executors"
-experimental                 "The comma separated set of experimental features to enable"
```

Note: exit will be called on bad parameter passing.

### CODE REVIEW

File: libs/settings/src/setting_collection.cpp [hunk 1]
```diff
@@ -103,8 +106,36 @@ void SettingCollection::UpdateFromArgs(int argc, char **argv)
...
+
+    for (auto const &setting : settings_)
+    {
+      oss << "\n-" << setting->name() << std::setw(int(30 - setting->name().size())) << " \""
+          << setting->description() << "\"";
+    }
+
+    FETCH_LOG_INFO(LOGGING_NAME, oss.str());
+
+    exit(1);
```
Comment (wkaluza) [inline libs/settings/src/setting_collection.cpp]: To improve testability, `std::exit` (or a wrapper) should be injected so it can be mocked out.

**Review by wkaluza — CHANGES_REQUESTED** [review]

File: libs/settings/src/setting_collection.cpp [hunk 2]
```diff
@@ -134,7 +134,7 @@ void SettingCollection::UpdateFromArgs(int argc, char **argv)
 
     FETCH_LOG_INFO(LOGGING_NAME, oss.str());
 
-    exit(1);
+    on_fail();
```
Comment (wkaluza) [inline libs/settings/src/setting_collection.cpp]: This needs a null check, I think.

**Review by wkaluza — CHANGES_REQUESTED** [review]

File: libs/settings/src/setting_collection.cpp (hunk 2 above)
Comment (n-hutton) [reply libs/settings/src/setting_collection.cpp]: Thanks, should be done now

**Review by wkaluza — DISMISSED** [review]

File: apps/constellation/settings.cpp [hunk 3]
```diff
@@ -89,7 +87,7 @@ Settings::Settings()
 bool Settings::Update(int argc, char **argv)
 {
   UpdateFromEnv("CONSTELLATION_");
-  UpdateFromArgs(argc, argv);
+  UpdateFromArgs(argc, argv, []() { std::exit(1); });
```
Comment (ejfitzgerald) [inline apps/constellation/settings.cpp]: My preference would be to get update from args to return a `bool` and handle this exit logic from here

File: libs/core/include/core/commandline/parameter_parser.hpp [hunk 4]
```diff
@@ -52,7 +52,10 @@ class ParamsParser
 
   std::string GetParam(std::string const &key, std::string const &default_value) const;
 
+  std::map<std::string, std::string> &params();
```
Comment (ejfitzgerald) [inline libs/core/include/core/commandline/parameter_parser.hpp]:
```suggestion
  std::map<std::string, std::string> const &params() const;
```

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]

**Review by ejfitzgerald — APPROVED** [review]

**Pull request merged**